### PR TITLE
Cp 3422/improvement/sort notes by date

### PR DESCRIPTION
**Pull request by alexholmes99**
Sorting notes by date

### CODE REVIEW

File: src/Components/Form.js [hunk 1]
```diff
@@ -34,71 +32,68 @@ function Form(props) {
...
-    setEmptyData((previousState) => {
-      return {
-        ...previousState,
-        [key]: value === "",
-      };
-    });
+  const handleBlur = (key) => {
+    if (key === "id") return;
+    if (formData[key] == "") {
+      document.getElementById(key).innerHTML = key + " required";
```
Comment (damonspencerroberts) [inline src/Components/Form.js]: why this change?

File: src/Components/Form.js [hunk 2]
```diff
@@ -34,71 +32,68 @@ function Form(props) {
...
-  const handleBlur = (value, key) => {
-    setEmptyData((previousState) => {
-      return {
-        ...previousState,
-        [key]: value === "",
-      };
-    });
+  const handleBlur = (key) => {
+    if (key === "id") return;
+    if (formData[key] == "") {
```
Comment (damonspencerroberts) [inline src/Components/Form.js]: triple equals here

File: src/Components/Form.js [hunk 3]
```diff
@@ -34,71 +32,68 @@ function Form(props) {
...
-    if (isValidated == true) {
-      const formDataCopy = { ...formData };
-      const notesCopy = [...notes];
-      notesCopy.push(formDataCopy);
-      dispatch(setNote(notesCopy));
+    const emptyValues = validateSubmit();
+    if (!emptyValues) {
+      let uniqueId = uuidv4();
+      let formDataCopy = { ...formData, id: uniqueId };
+      console.log(formDataCopy);
```
Comment (damonspencerroberts) [inline src/Components/Form.js]: remove

File: src/Components/Form.js [hunk 4]
```diff
@@ -34,71 +32,68 @@ function Form(props) {
...
-    const isValidated = validateSubmit();
-    if (isValidated == true) {
-      const formDataCopy = { ...formData };
-      const notesCopy = [...notes];
-      notesCopy.push(formDataCopy);
-      dispatch(setNote(notesCopy));
+    const emptyValues = validateSubmit();
+    if (!emptyValues) {
+      let uniqueId = uuidv4();
+      let formDataCopy = { ...formData, id: uniqueId };
```
Comment (damonspencerroberts) [inline src/Components/Form.js]: why are these `let` and not `const`?

File: src/Components/Form.js [hunk 5]
```diff
@@ -34,71 +32,68 @@ function Form(props) {
...
+      console.log(formDataCopy);
+      dispatch(setNote(formDataCopy));
     }
   };
 
   return (
     <OuterForm onSubmit={handleSubmit}>
       <FormContent>
-        {emptyData.name && <ErrorMessage>Name required</ErrorMessage>}
+        {<ErrorMessage id="name"></ErrorMessage>}
```
Comment (damonspencerroberts) [inline src/Components/Form.js]: confused on the change here

File: src/Components/Form.js [hunk 6]
```diff
@@ -14,16 +14,14 @@ function Form(props) {
...
     return state.noteReducer.value;
   });
 
   const dispatch = useDispatch();
 
-  const [emptyData, setEmptyData] = useState({
-    name: false,
-    date: false,
-    note: false,
+  const [modifiers, setModifiers] = useState({
```
Comment (damonspencerroberts) [inline src/Components/Form.js]: where is this being used?

File: src/Components/Notes.js [hunk 7]
```diff
@@ -1,22 +1,62 @@
...
+  const sortNotes = (e) => {
+    e.preventDefault();
+    setSorted(!sorted);
+  };
+
+  const noteMap = () => {
+    let sortedData = [...notes];
+    if (sorted) {
+      sortedData = sortedData.sort((a, b) => (a.date > b.date ? 1 : -1));
+    } else {
```
Comment (damonspencerroberts) [inline src/Components/Notes.js]: no need for an else here

File: src/Components/Notes.js [hunk 8]
```diff
@@ -1,22 +1,62 @@
...
+            fontsize: "24px",
+            width: "10%",
+            float: "left",
+            background: "none",
+          }}
+        >
+          {sorted ? "Undo" : "Sort Notes"}
+        </Button>
+      ) : null}
+      {notes.length === 0 ? (
```
Comment (damonspencerroberts) [inline src/Components/Notes.js]: You can use `{notes.length === 0 && <NoNotes> Notes will be displayed here </NoNotes>}` here

File: src/Components/Notes.js [hunk 9]
```diff
@@ -1,22 +1,62 @@
...
-      {notes.map((n) => {
-        return (
-          <StickyNote>
-            <NoteDate>{n.date}</NoteDate>
-            <NoteName>{n.name}</NoteName>
-            <NoteContent>{n.note}</NoteContent>
-          </StickyNote>
-        );
-      })}
+      {notes.length != 0 ? (
```
Comment (damonspencerroberts) [inline src/Components/Notes.js]: `!==`

Comment (damonspencerroberts) [reply src/Components/Notes.js]: you can use 
```
{notes.length !== 0 && <Button
          type="Sort"
          value="Sort"
          id="Sort"
          onClick={sortNotes}
          cssOptions={{
            border: "none",
            opacity: "0.8",
            textdecor: "underline",
            fontsize: "24px",
            width: "10%",
            float: "left",
            background: "none",
          }}
        >
          {sorted ? "Undo" : "Sort Notes"}
        </Button>
```

File: src/Components/Notes.js (hunk 8 above)
Comment (damonspencerroberts) [reply src/Components/Notes.js]: Instead of two conditions here, why dont you have a ternary:

```
{notes.length === 0 ? (
<NoNotes> Notes will be displayed here </NoNotes>
) : (
<Button
          type="Sort"
          value="Sort"
          id="Sort"
          onClick={sortNotes}
          cssOptions={{
            border: "none",
            opacity: "0.8",
            textdecor: "underline",
            fontsize: "24px",
            width: "10%",
            float: "left",
            background: "none",
          }}
        >
          {sorted ? "Undo" : "Sort Notes"}
        </Button>
)}
```

File: src/Redux/Reducers/noteReducer.js [hunk 10]
```diff
@@ -7,7 +7,10 @@ export const initialState = {
 export default function (state = initialState, action) {
   switch (action.type) {
     case actions.SET_NOTE:
-      return { ...state, value: action.value };
+      let x = state.value;
```
Comment (damonspencerroberts) [inline src/Redux/Reducers/noteReducer.js]: use a better variable name than `x`

Comment (damonspencerroberts) [reply src/Redux/Reducers/noteReducer.js]: no need to use `let` here you can use `const`

File: src/Redux/Reducers/noteReducer.js [hunk 11]
```diff
@@ -7,7 +7,10 @@ export const initialState = {
...
+      let dataArray = [...x];
```
Comment (damonspencerroberts) [inline src/Redux/Reducers/noteReducer.js]: same here you can use `const`

**Review by damonspencerroberts — CHANGES_REQUESTED** [review]
Couple of changes to make. Make sure you know the difference between `let` and `const`. `let` is really only used if you are going to modify the variable assignment. When you are modifying an array or an object, you can still use `const`.
Example:
**Use Let**
```
let a = 1
if (a > 0) {
   a = 2
}
```
**Use const**
```
const a = []
if (a.length === 0) {
   a.push(1)
}
```

File: src/Components/Form.js (hunk 1 above)
Comment (alexholmes99) [reply src/Components/Form.js]: Was told to figure out a method w/o using emptyData state

File: src/Components/Form.js (hunk 5 above)
Comment (alexholmes99) [reply src/Components/Form.js]: Was told to do it w/o emptyData state

File: src/Components/Form.js (hunk 1 above)
Comment (damonspencerroberts) [reply src/Components/Form.js]: @alexholmes99 sounds good, try doing this with `useRef` instead let me know if you need help. Check here for documentation... [here](https://reactjs.org/docs/hooks-reference.html#useref)

File: src/Redux/Reducers/noteReducer.js [hunk 12]
```diff
@@ -7,8 +7,8 @@ export const initialState = {
 export default function (state = initialState, action) {
   switch (action.type) {
     case actions.SET_NOTE:
-      let x = state.value;
-      let dataArray = [...x];
+      const newData = state.value;
```
Comment (damonspencerroberts) [inline src/Redux/Reducers/noteReducer.js]: Hate to be annoying here but maybe this should be `existingData` as it is coming from the state and wouldn't be new data

**Review by damonspencerroberts — CHANGES_REQUESTED** [review]
one small comment and reply to message above

File: src/Components/Form.js [hunk 13]
```diff
@@ -14,17 +14,13 @@ function Form(props) {
     note: "",
   });
 
-  const notes = useSelector((state) => {
+  let notes = useSelector((state) => {
```
Comment (damonspencerroberts) [inline src/Components/Form.js:17]: should probably be `const` unless you are changing it somewhere

File: src/Components/Form.js [hunk 14]
```diff
@@ -34,71 +30,76 @@ function Form(props) {
...
+      const formDataCopy = { ...formData, id: uniqueId };
+      dispatch(setNote(formDataCopy));
     }
   };
 
   return (
     <OuterForm onSubmit={handleSubmit}>
       <FormContent>
-        {emptyData.name && <ErrorMessage>Name required</ErrorMessage>}
+        {
```
Comment (damonspencerroberts) [inline src/Components/Form.js]: you can remove curly brackets from all error message only necessary when writing some logic, functions, variables, etc.

File: src/Components/Notes.js [hunk 15]
```diff
@@ -1,22 +1,61 @@
...
 import { useSelector } from "react-redux";
 import styled from "styled-components";
+import Button from "./Button";
 
 function Notes() {
   const notes = useSelector((state) => {
     return state.noteReducer.value;
   });
+
+  const [sorted, setSorted] = useState(false);
```
Comment (damonspencerroberts) [inline src/Components/Notes.js]: totally up to you, not necessary, but might be good practice trying to set this with Redux. Completely up to you.

File: src/Components/Notes.js [hunk 16]
```diff
@@ -1,22 +1,61 @@
...
+
+  const sortNotes = (e) => {
+    e.preventDefault();
+    setSorted(!sorted);
+  };
+
+  const noteMap = () => {
+    const sortedData = [...notes];
```
Comment (damonspencerroberts) [inline src/Components/Notes.js]: in this case you would use `let` because you are possibly resetting it in this line.

```
sortedData = sortedData.sort((a, b) => (a.date > b.date ? 1 : -1));
```

**Review by damonspencerroberts — CHANGES_REQUESTED** [review]
Couple of small things.

File: src/Components/Form.js (hunk 13 above)
Comment (alexholmes99) [reply src/Components/Form.js:17]: Changing it to const causes an error when sorting then notes: Uncaught TypeError: Assignment to constant variable.
